### ISSUE TITLE
[Bug]: Model mixin conflicts with ModernFix dynamic resources

### ISSUE THREAD

**Issue by embeddedt** [Fuzss/diagonal-blocks#47]
### Mod Loader (Required)

Forge

### Minecraft Version (Required)

All since 1.18.2

### Mod Version (Required)

N/A

### Notes (Required)

Diagonal Fences (and Diagonal Windows) inject at the HEAD of `uploadTextures` to replace fence models. This is not compatible with ModernFix dynamic resources as it makes many changes to the model loading implementation.

It's best to use platform-provided model wrapping hooks (such as Model Loading API v1 on Fabric, or `ModelBakeEvent` on Forge) but if these are not suitable/available, I would suggest use of a `ModifyVariable` on `cacheAndQueueDependencies` instead, like so:

```java
@ModifyVariable(method = "cacheAndQueueDependencies", at = @At("HEAD"), argsOnly = true)
private UnbakedModel replaceFenceModel(UnbakedModel model, ResourceLocation location) {
    
}
```

and then returning whatever model instance you want to get used. This avoids relying on the assumption that all models are going to be baked after `onPrepareModelBaking` is called.

The current solution I have been using is a mixin into Diagonal Fences internals but unfortunately the implementation has been changed, breaking the mixin, and this solution also isn't feasible to duplicate for a second "diagonal blocks" mod.

**Comment by Fuzss** [comment]
So the custom mixin was sort of necessary as I need to modify unbaked models (to find the correct multi part conditions which is not possible for baked models where the conditions have been turned into predicates) which at least on Forge isn't very straight forward since the only event for modifying models runs after models have been baked.

But I reworked the implementation now to utilize the new model api on Fabric and the existing `ModelEvent$ModifyBakingResult` event on Forge (at the cost of having to rebake the models). Here are the relevant parts of the implementation:
- [Diagonal Blocks: Event usage](https://github.com/Fuzss/diagonalblocks/blob/a09efec6c8ecccb450aedf3bd27dae55e5df4d0a/1.20/Common/src/main/java/fuzs/diagonalblocks/client/handler/DiagonalModelHandler.java#L34)
- [Puzzles Lib (Common): Event class](https://github.com/Fuzss/puzzleslib/blob/d3fc00fc74dc0907e432e3a71416462c19937a21/1.20/Common/src/main/java/fuzs/puzzleslib/api/client/event/v1/ModelEvents.java#L50)
- [Puzzles Lib (Fabric): Implementation for common event](https://github.com/Fuzss/puzzleslib/blob/d3fc00fc74dc0907e432e3a71416462c19937a21/1.20/Fabric/src/main/java/fuzs/puzzleslib/impl/client/event/FabricClientEventInvokers.java#L361)
- [Puzzles Lib (Forge): Implementation for common event](https://github.com/Fuzss/puzzleslib/blob/d3fc00fc74dc0907e432e3a71416462c19937a21/1.20/Forge/src/main/java/fuzs/puzzleslib/impl/client/event/ForgeClientEventInvokers.java#L421)

Is this enough to fix compatibility with ModernFix?
From my simple tests at least there is no crash and diagonal connections are visible. That's what I heard was causing issues for players although I never was able to reproduce either with the various mod versions of both mods I tested. Not sure if there was some workaround on your end in the meantime?

**Comment by embeddedt** [comment]
You cannot iterate over the entry set [here](https://github.com/Fuzss/puzzleslib/blob/d3fc00fc74dc0907e432e3a71416462c19937a21/1.20/Forge/src/main/java/fuzs/puzzleslib/impl/client/event/ForgeClientEventInvokers.java#L442), as that map will appear empty to iterators when dynamic resources is enabled. This is because 99% of the models are not already loaded when the event fires.

Instead I would recommend iterating over the block registry, filtering for fence blocks, and calculate the appropriate `ModelResourceLocation` for each fence state. Then, you can iterate over that list of locations and retrieve & set things in the `models` map as needed.

In my tests with ModernFix 5.8.1 and Diagonal Fences 8.1.1, fences currently just render as stone blocks when dynamic resources is enabled (presumably due to what I described above).

**Comment by Fuzss** [comment]
Yeah, makes sense to behave like that. Before the model event all diagonal blocks get a stone block model to prevent the bakery from logging a missing model.

I mocked up a quick implementation that collects all the top level model locations on its own and iterates over those, with the current model coming from a supplier. I really want an implementation like Fabric that receives every model (or top level model at least). This approach works with dynamic resources enabled.

A few quick questions:
1. Is it safe to assume the missing model is present in the baked top level models map provided by Forge's `ModelEvent$BakingCompleted`?
2. Is there a downside to using a [baked model cache](https://github.com/Fuzss/puzzleslib/blob/d3fc00fc74dc0907e432e3a71416462c19937a21/1.20/Forge/src/main/java/fuzs/puzzleslib/impl/client/event/ForgeClientEventInvokers.java#L434) storing the baked models with their unbaked model as key? Vanilla uses this custom `BakedCacheKey` consisting of ResourceLocation, Transformation and a boolean. But since the resource location is usually a model location it is different for every block state. So vanilla bakes an unbaked model again for every possible block state, although the unbaked model is the same and so is the transformation and uv lock property. I implemented my cache with the unbaked model as key and haven't noticed any issue. Are you aware of any downsides?

**Comment by embeddedt** [comment]
> Is it safe to assume the missing model is present in the baked top level models map provided by Forge's `ModelEvent$BakingCompleted`?

Yes, retrieving the missing model by name should always work. A lot of things would break otherwise. In fact, retrieving any model by name should work, you just can't iterate over the contents of the map.

> I implemented my cache with the unbaked model as key and haven't noticed any issue. Are you aware of any downsides?

I was under the impression vanilla also generates a separate unbaked model for each block state, and FerriteCore was the one which deduplicated that for multipart models like fences. But if this works correctly I don't see a reason to change it.

**Comment by Fuzss** [comment]
Great, then I'll just polish that a little more. Thanks for the quick response!

And no, vanilla definitely only uses a single unbaked model per block. My previous Mixin implementation for modifying unbaked models was [based on that fact](https://github.com/Fuzss/diagonalblocks/blob/fa46f8d0a7667f5783bd5546467bddd3b95e8236/1.20/Common/src/main/java/fuzs/diagonalfences/client/model/MultipartAppender.java#L41), I only needed to find the unbaked model for the default block state.

**Comment by Fuzss** [comment]
Fixed now in the latest version. I just hope I didn't break any of your optimizations. But looking good so far. 😅

**Comment by Fuzss** [comment]
Should probably leave this open until the new version makes its way to 1.18/1.19.
